### PR TITLE
Set the clean up policy to compact for non-windowed tables.

### PR DESCRIPTION
**Pull request by hjafarpour**
Currently we don't set the clean up policy for sink topics and it uses the default which is `delete`. For the non-window tables we should use `compact` since these are changelog topics.
For the windowed tables we still use `delete` clean up policy since we would prefer to not keep the windowed table forever!

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java [hunk 1]
```diff
@@ -133,7 +133,11 @@ public SchemaKStream buildStream(
     );
 
     final KsqlStructuredDataOutputNode noRowKey = outputNodeBuilder.build();
-    createSinkTopic(noRowKey.getKafkaTopicName(), ksqlConfig, kafkaTopicClient);
+    createSinkTopic(noRowKey.getKafkaTopicName(),
+                    ksqlConfig,
+                    kafkaTopicClient,
+                    ((result instanceof SchemaKTable)
+                     && !((SchemaKTable) result).isWindowed()));
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java]: I would move this boolean to a helper method named `shouldCompact`. Makes the intent much clearer.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java [hunk 2]
```diff
@@ -129,6 +137,12 @@ public boolean isTopicExists(final String topic) {
     }
   }
 
+  @Override
+  public DescribeConfigsResult describeConfigs(String topicName) {
+    return adminClient.describeConfigs(
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java]: This should use the `RetryHelper.executeWithRetries` method. Otherwise the command will fail for even retriable exceptions.

File: ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java [hunk 3]
```diff
@@ -187,6 +203,44 @@ public void testSinkProperties() throws Exception {
     assertThat(
         topicClient.describeTopics(ImmutableList.of(streamName)).get(streamName).partitions(),
         hasSize(3));
+    Map<ConfigResource, Config> configResourceConfigMap = topicClient.describeConfigs
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java]: This block is used 3 times. Should be moved to a utility method named `verifyCleanupPolicy(String policy)`. Then you can get the property from the cluster and verify that it matches the expectation.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 4]
```diff
@@ -92,7 +93,7 @@ public void before() {
     props.put(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY, 4);
     props.put(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, (short)3);
     createOutputNode(props);
-    topicClient.createTopic(eq("output"), anyInt(), anyShort());
+    topicClient.createTopic(eq("output"), anyInt(), anyShort(), anyBoolean());
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: We should add a unit test here that verifies that non windowed KTables actually are being created with `cleanup.polic=compact`.

File: ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java [hunk 5]
```diff
@@ -316,12 +317,16 @@ private DescribeClusterResult getDescribeClusterResult() {
     return describeClusterResult;
   }
 
-  private DescribeConfigsResult getDescribeConfigsResult() {
+  private DescribeConfigsResult getDescribeConfigsResult(boolean isCompacted) {
     DescribeConfigsResult describeConfigsResult = mock(DescribeConfigsResult.class);
-    ConfigEntry configEntry = new ConfigEntry("delete.topic.enable", "true");
+    ConfigEntry configEntryDeleteEnable = new ConfigEntry("delete.topic.enable", "true");
+    ConfigEntry configEntryIsCompacted = new ConfigEntry("cleanup.policy",
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java]: Is this returned config entry actually used anywhere in admin client? The only usage of describeConfigs I see is to get the broker's delete.topic.enable config. If cleanup.policy is not read anywhere can we just get rid of it and the isCompacted parameter to this function.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java (hunk 2 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java]: Changed the method to use RetryHelper and return the clean up policy only.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java (hunk 4 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: Added a unit test.

File: ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java (hunk 5 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java]: No it was only used in tests. Removed it.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java [hunk 6]
```diff
@@ -74,11 +73,12 @@ void createTopic(
   Map<String, TopicDescription> describeTopics(Collection<String> topicNames);
 
   /**
-   * [warn] synchronous call to get the response
+   * [warn] Asynchronous call to get the response
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: This is a synchronous call. It blocks until it gets a successful response or will throw an exception if it didn't get a successful response after 5 retries.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java [hunk 7]
```diff
@@ -74,11 +73,12 @@ void createTopic(
...
    *
-   * @param topicName topicName to describe
+   * @param topicName topicNames to describe
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: Why make this plural. It is still a single topic.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java [hunk 8]
```diff
@@ -74,11 +73,12 @@ void createTopic(
...
    */
-  DescribeConfigsResult describeConfigs(String topicName);
+  public String getTopicCleanupPolicy(String topicName);
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: Would be better if this was an enum with three values: `delete`, `compact`, `compact+delete`. Would be better than each user of the method relying on the actual string literals.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java (hunk 6 above)
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: this call looks synchronous to me

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 9]
```diff
@@ -189,4 +191,47 @@ private SchemaKStream buildStream() {
...
+                                                  schema,
+                                                  schema.field("timestamp"),
+                                                  schema.field("key"),
+                                                  new KsqlTopic("output", "output", new KsqlJsonTopicSerDe()),
+                                                  "output",
+                                                  props,
+                                                  Optional.empty());
+
+    StreamsBuilder streamsBuilder = new StreamsBuilder();
+    SchemaKTable schemaKTable = (SchemaKTable) outputNode.buildStream(streamsBuilder,
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: We should first have a check `assertThat(scheamKstream, instanceof(schemaKTable))` and then do the cast.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 10]
```diff
@@ -189,4 +191,47 @@ private SchemaKStream buildStream() {
...
+                                                                      ksqlConfig,
+                                                                      topicClientForNonWindowTable,
+                                                                      new MetastoreUtil(),
+                                                                      new FunctionRegistry(),
+                                                                      new HashMap<>(),
+                                                                      new MockSchemaRegistryClient());
+    topicClientForNonWindowTable.createTopic(eq("output"), anyInt(), anyShort(), eq(true));
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: Would be better to have `4` and`3` instead fo `anyInt()` and `anyShort()`.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 11]
```diff
@@ -189,4 +191,47 @@ private SchemaKStream buildStream() {
...
+    EasyMock.expectLastCall();
+    EasyMock.replay(topicClientForNonWindowTable);
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: This seems off.

We should be doing: 

```java
topicClientForNonWindowTable.createTopic(...)
Easymock.expectLastCall();
EasyMock.replay(topicClientForNonWindownTable);
outputNode.buildStream(...);
EasyMock.verify();
```

This way we are actually checking that the right call was made to the topic client when trying to create the sink topic.

Also, we should add tests which check that the compaction policy is delete for streams and windowed tables.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java (hunk 10 above)
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: Shouldn't this plus the next 2 lines come before buildStream?

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java (hunk 7 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: This calls `AdminClient.describeTopics` which gets a collection of topics.

Comment (apurvam) [reply ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClient.java]: Yes. But this method just takes a single topic and passes a list with a single element and also assumes the output contains a single pair. 

So it is more appropriate for javadoc for this method to be singular rather than plural.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java (hunk 11 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: Made changes and added tests for both windowed table and stream.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java (hunk 10 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: Yes, indeed this needed more changes.

File: ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java [hunk 12]
```diff
@@ -187,6 +203,24 @@ public void testSinkProperties() throws Exception {
...
+                                             + "FROM %s;",
+                                             tableName, usersTable);
+    executePersistentQuery(queryString);
+
+    TestUtils.waitForCondition(
+        () -> topicClient.isTopicExists(tableName),
+        "Wait for async topic creation"
+    );
+
+    assertThat(topicClient.getTopicCleanupPolicy(tableName), equalTo("compact"));
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java]: here and elsewhere we should be checking equivalence to `TopicCleanupPolicy.COMPACT` etc.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 13]
```diff
@@ -188,4 +192,90 @@ private SchemaKStream buildStream() {
...
 
+  @Test
+  public void shouldCreateSinkWithCorrectCleanupPolicyNonWindowedTable() {
+    KafkaTopicClient topicClientForNonWindowTable = EasyMock.mock(KafkaTopicClient.class);
+    KsqlStructuredDataOutputNode outputNode = getKsqlStructuredDataOutputNode(false);
+    StreamsBuilder streamsBuilder = new StreamsBuilder();
+    topicClientForNonWindowTable.createTopic("output", 4, (short) 3, true);
+    EasyMock.replay(topicClientForNonWindowTable);
+    SchemaKStream schemaKStream = outputNode.buildStream(streamsBuilder,
+                                                                      ksqlConfig,
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: indentation.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java [hunk 14]
```diff
@@ -188,4 +192,90 @@ private SchemaKStream buildStream() {
...
+                      schema.field("key"),
+                      schema.field("timestamp"),
+                      new KsqlTopic("input", "input",
+                                    new KsqlJsonTopicSerDe()),
+                      "TableStateStore",
+                      isWindowed),
+        schema);
+
+    return new KsqlStructuredDataOutputNode(new PlanNodeId("0"),
+                                                  tableSourceNode,
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java]: indentation.

File: ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java [hunk 15]
```diff
@@ -316,12 +317,13 @@ private DescribeClusterResult getDescribeClusterResult() {
     return describeClusterResult;
   }
 
-  private DescribeConfigsResult getDescribeConfigsResult() {
+  private DescribeConfigsResult getDescribeConfigsResult(boolean isCompacted) {
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java]: Why this change? This variable doesn't seem to be used at all and really doesn't need to be changed relative to the other changes in this PR.

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java [hunk 16]
```diff
@@ -64,24 +66,31 @@ public KafkaTopicClientImpl(final AdminClient adminClient) {
   public void createTopic(
       final String topic,
       final int numPartitions,
-      final short replicatonFactor
+      final short replicatonFactor,
+      boolean isCompacted
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java]: We should use the TopicCleanupPolicy here instead of isCompacted. Also, wouldn't it be better to use compact+delete for topics backing windowed aggregations?

Both of these changes can be made in a later PR targeting master

File: ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java [hunk 17]
```diff
@@ -129,6 +138,45 @@ public boolean isTopicExists(final String topic) {
...
+    }
+    if (configMap == null) {
+      throw new KsqlException("Could not get the topic configs for : " + topicName);
+    }
+    Object[] configValues = configMap.values().stream().findFirst().get()
+        .entries()
+        .stream()
+        .filter(configEntry -> configEntry.name().equalsIgnoreCase("cleanup.policy"))
+        .toArray();
+    if (configValues == null || configValues.length ==0) {
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/util/KafkaTopicClientImpl.java]: nit: == 0

File: ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java (hunk 15 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/util/KafkaTopicClientTest.java]: Yes, it was left from previous changes. Removed it.
